Test on GitHub Actions

hugovk opened the pull request:
Travis CI is being turned off for the pytest-dev org ([internal discussion](https://github.com/orgs/pytest-dev/teams/contributors/discussions/5)), due to the recent security handling ([1](https://twitter.com/peter_szilagyi/status/1437646118700175360), [2](https://arstechnica.com/information-technology/2021/09/travis-ci-flaw-exposed-secrets-for-thousands-of-open-source-projects/)).

This PR replaces Travis CI with GitHub Actions. It mostly replicates the old Travis workflow, here's the main differences:

* Do the linting/examples/test stuff in parallel, not in serial. We now have 20 parallel jobs with GHA, so let's use them.
* I left out these global env vars, are they needed?

```yml
env:
  global:
    - LD_PRELOAD=/lib/x86_64-linux-gnu/libSegFault.so
    - SEGFAULT_SIGNALS=all
```
* I added caching to speed up builds and ease the load on PyPI
* No email sending with GHA on failure
* Don't test Python 2.7 and 3.5, support was already dropped in https://github.com/pytest-dev/pytest-cov/pull/488

---

# Question

Do we need to test on all these?

```yml
            "pytest46-xdist127",
            "pytest46-xdist133",
            "pytest54-xdist133",
            "pytest62-xdist202",
```

(Ignoring the few jobs removed from the matrix:)

The `test` matrix is 5 Python x 4 pytest+xdist = 20 jobs (soon 6x4=24 with Python 3.10).

I'd suggest testing on just the last two pytest+xdist, and then consider testing also macOS on GHA:

2 OS x 5 Python x 2 pytest+xdist = 20 jobs (2x6x2=24 with Python 3.10).

hugovk commented:
Actions will need enabling for this repo. Please can someone turn it on? It's probably at https://github.com/pytest-dev/actions or maybe https://github.com/pytest-dev/pytest-cov/settings/actions.

Example runs:

* lint https://github.com/hugovk/pytest-cov/actions/runs/1254004204
* examples https://github.com/hugovk/pytest-cov/actions/runs/1254004205
* test https://github.com/hugovk/pytest-cov/actions/runs/1254004208

hugovk commented:
> Actions will need enabling for this repo. Please can someone turn it on? It's probably at [pytest-dev/actions](https://github.com/pytest-dev/actions) or maybe [pytest-dev/pytest-cov/settings/actions](https://github.com/pytest-dev/pytest-cov/settings/actions).

Correction: nothing needs enabling in the repo, it just needs the workflow merged (re: https://github.com/pytest-dev/nose2pytest/pull/18#issuecomment-923069805).

graingert commented:
> Correction: nothing needs enabling in the repo, it just needs the workflow merged (re: pytest-dev/nose2pytest#18 (comment)).

do you want to make a "hello world" workflow so we can see this run in PRs?

hugovk commented:
Sure, please see PR https://github.com/pytest-dev/pytest-cov/pull/495!

graingert commented:
do we need appveyor anymore?

graingert commented:
can you add an ["all good" job](https://github.community/t/is-it-possible-to-require-all-github-actions-tasks-to-pass-without-enumerating-them/117957/4?u=graingert) that we can add as a "required build",

graingert commented on MANIFEST.in:
```diff
@@ -6,6 +6,7 @@ prune examples/*/*/htmlcov
 prune examples/adhoc-layout/*.egg-info
 prune examples/src-layout/src/*.egg-info
 
+graft .github/workflows
```
is there a reason for publishing the github workflow to pypi? they both use `actions/checkout@v2` which won't work without a live git repo and can't work from the sdist

hugovk replied:
Yes, the reason is `.travis.yml .appveyor.yml` were/is included below, so I was aiming for parity.

Can remove if you like?

graingert replied:
I don't know the best option here is.

I think using the "wheel_from_sdist", and then using that wheel in tox is the state of the art for ensuring your MANIFEST is correct https://github.com/pypa/build/issues/257

hugovk replied:
No strong opinion here, although I use setuptools-scm elsewhere and it makes things easy. Something for a followup issue?

ionelmc replied:
I think it's fair to include it in the sdist, considering travis stuff was previously included, and it would be useful to have some evidence in the sdist of how that particular release was supposed to be tested.

graingert commented on ci/templates/.travis.yml:
```diff
@@ -1,61 +0,0 @@
-cache: false
-env:
-  global:
-    - LD_PRELOAD=/lib/x86_64-linux-gnu/libSegFault.so
-    - SEGFAULT_SIGNALS=all
```
is this libSegFault/SEGFAULT_SIGNALS=all something important that needs to be ported to GHA?

graingert replied:
ah you asked this in the PR description:

> * I left out these global env vars, are they needed?
> 
> ```yml
> env:
>   global:
>     - LD_PRELOAD=/lib/x86_64-linux-gnu/libSegFault.so
>     - SEGFAULT_SIGNALS=all
> ```

graingert replied:
seems like it was introduced here https://github.com/pytest-dev/pytest-cov/pull/58/files#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R6

graingert commented on docs/reporting.rst:
```diff
@@ -71,4 +71,4 @@ The final report option can also suppress printing to the terminal::
 This mode can be especially useful on continuous integration servers, where a coverage file
 is needed for subsequent processing, but no local report needs to be viewed. For example,
-tests run on Travis-CI could produce a .coverage file for use with Coveralls.
+tests run on GitHub Actions could produce a .coverage file for use with Coveralls.
```
this is a CI recommendation and coverage provider recommendation, eg GitLab has integrated support for coverage files (and recommend pytest-cov!) https://docs.gitlab.com/ee/user/project/merge_requests/test_coverage_visualization.html#python-example

graingert commented on examples/adhoc-layout/example/__init__.py:
```diff
@@ -1,12 +1,2 @@
-else:
-    def add(a, b):
-        return a + b
+def add(a, b):
+    return a + b
```
I think this is testing conditional collection and merging - perhaps something like:

```
import platform

if platform.python_implementation() == "PyPy":
    def add(a, b):
        return a + b

else:
    def add(a, b):
        return a + b
```

and then run this example on pypy3.7 and cpython3.9 ?

hugovk replied:
Yep, updated!

graingert commented on examples/adhoc-layout/tox.ini:
```diff
@@ -19,7 +19,7 @@ deps =
+    report: py38
 
 # note that this is necessary to prevent the tests importing the code from your badly laid project
 changedir = tests
```
afaik this is fixed by coverage's `source_pkg` flag

graingert replied:
https://github.com/pytest-dev/pytest-cov/issues/496

graingert commented on examples/src-layout/src/example/__init__.py:
```diff
@@ -1,12 +1,2 @@
-else:
-    def add(a, b):
-        return a + b
+def add(a, b):
```
same again here:

```python
import platform

# test merging multiple tox runs with a platform
# based branch
if platform.python_implementation() == "PyPy":
    def add(a, b):
        return a + b

else:
    def add(a, b):
        return a + b
```

graingert commented on .github/workflows/test.yml:
```diff
@@ -11,7 +11,26 @@ jobs:
       fail-fast: false
       matrix:
-        python-version: ["3.9"]
+        python-version: ["pypy3", "3.6", "3.7", "3.8", "3.9"]
```
I think `pypy3` is an alias for the old `pypy-3.6`

```suggestion
        python-version: ["pypy-3.6", "pypy-3.7", "3.6", "3.7", "3.8", "3.9"]
```

graingert replied:
yep: https://github.com/pytest-dev/pytest-cov/runs/3660088990?check_suite_focus=true#step:3:7

graingert replied:
hmm it arrived from pytest-cover here https://github.com/pytest-dev/pytest-cov/pull/58/files#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R6

hugovk commented:
> do we need appveyor anymore?

Maybe not, I had a very quick look, there's possibly stuff going on in `.\ci\appveyor-with-compiler.cmd` that's no longer needed. But for this PR I want to keep the scope to replacing Travis CI as it's about to be turned off. And AppVeyor is still ticking over fine.

hugovk commented:
> can you add an ["all good" job](https://github.community/t/is-it-possible-to-require-all-github-actions-tasks-to-pass-without-enumerating-them/117957/4?u=graingert) that we can add as a "required build",

Done!

hugovk commented:
"All good" is now all good!

* https://github.com/pytest-dev/pytest-cov/pull/494/checks?check_run_id=3660738703

I think I've addressed all the review comments, ready for re-review!

graingert commented:
@hugovk btw there's a pytest-cov discord channel https://discord.gg/AkvNx9gp2T
